clang/include/clang/AST/DeclCXX.h

```diff
@@ -1392,7 +1392,8 @@ class CXXRecordDecl : public RecordDecl {
   bool allowConstDefaultInit() const {
     return !data().HasUninitializedFields ||
            !(data().HasDefaultedDefaultConstructor ||
-             needsImplicitDefaultConstructor());
+             needsImplicitDefaultConstructor()) ||
+           hasInClassInitializer();
```

Comment:
Based on the comment in the bug report I think this needs to be `(isUnion() && hasInClassInitializer())`

Reply:
What happens for:

```
struct A {
  union {
    struct {
      int n = 5;
      int m;
    };
  };
};
const A a;
```

This is a language extension so we get to choose its semantics, but I think the most logical choice is to reject because the union member is not *fully* initialized.

Reply:
> This is a language extension so we get to choose its semantics, but I think the most logical choice is to reject because the union member is not _fully_ initialized.

At the moment, this gets accepted.

Reply:
The two tests that were broken:

```c++
struct some_init {
  int a = 0;
  int b;
  int c = 0;
};

void constobjs() {
  // ...
  const some_init si;
  // ...
}
```

```c++
struct Ints2 {
  int a = 10;
  int b;
};
constexpr Ints2 ints22;
```

Reply:
@zygoloid 
Is there an existing method for checking nested structures for cases like you mentioned above?

Reply:
I don't think so. Maybe we can change the behavior of `HasUninitializedFields` -- it only exists to support this check, so we can change it to do whatever this check needs. What I'm thinking is:

* If the class is not a union, then proceed as we currently do: start it off as `false` and set it to `true` if we find an uninitialized field.
* If the class is a union, then do the opposite: start it off as `true` and set it to `false` if we find an initialized field.

(So for a non-union, the field is true if all fields are const default initializable, and for a union, the field is true if *any* field is const default initializable.)

Then I think the only change we'd need here would be to also treat empty unions as const default initializable.

Reply:
Hi @zygoloid 
I apologize if it's a silly question, I'm new here and want to learn something from you.
@Rajveer100's commit has already fixed the test (GH95854.cpp).
> The two tests that were broken:

And these two mentioned testcases' errors are expected. The behavior of the fix is correct.
Then, what are you planning to do next?
Are you planning to improve the code by changing the behavior of `HasUninitializedFields`?